### PR TITLE
Serialize incr comp structures to file via fixed-size buffer

### PR DESCRIPTION
**Pull request by tgnottingham**
Reduce a large memory spike that happens during serialization by writing
the incr comp structures to file by way of a fixed-size buffer, rather
than an unbounded vector.

Effort was made to keep the instruction count close to that of the
previous implementation. However, buffered writing to a file inherently
has more overhead than writing to a vector, because each write may
result in a handleable error. To reduce this overhead, arrangements are
made so that each LEB128-encoded integer can be written to the buffer
with only one capacity and error check. Higher-level optimizations in
which entire composite structures can be written with one capacity and
error check are possible, but would require much more work.

The performance is mostly on par with the previous implementation, with
small to moderate instruction count regressions. The memory reduction is
significant, however, so it seems like a worth-while trade-off.

### CODE REVIEW

**Comment by tgnottingham** [comment]
Btw, if #80115 merges and this is rebased on and adapted for it, it should also reduce the impact on instruction count.

**Comment by tgnottingham** [comment]
Oops, forgot to update tests. :sweat_smile:

**Comment by jyn514** [comment]
I would be curious to see how much this helps with https://github.com/rust-lang/rust/issues/79103.

**Comment by tgnottingham** [comment]
The results are more extreme than what I see in local benchmarks. A 20% reduction turned into a 30% reduction, but the instruction counts are worse. I'll investigate.

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -1338,7 +1339,7 @@ impl<'tcx> TyCtxt<'tcx> {
 
     pub fn serialize_query_result_cache<E>(self, encoder: &mut E) -> Result<(), E::Error>
     where
-        E: ty::codec::OpaqueEncoder,
+        E: OpaqueEncoder,
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:1342]: This function is only ever called with `FileEncoder`. It could be made monomorphic.

Comment (cjgillot) [reply compiler/rustc_middle/src/ty/context.rs:1342]: `OnDiskCache::serialize` and `encode_query_results` could also.

File: compiler/rustc_middle/src/ty/codec.rs [hunk 2]
```diff
@@ -50,22 +50,6 @@ impl<'tcx, E: TyEncoder<'tcx>> EncodableWithShorthand<'tcx, E> for ty::Predicate
     }
 }
 
-pub trait OpaqueEncoder: Encoder {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/codec.rs:53]: Right now, the only use for `OpaqueEncoder` is to workaround a specialization bug which prevents from using `opaque::Encoder` everywhere in the `OnDiskCache`. `OpaqueEncoder` may rather belong to `ty::query::on_disk_cache`, with its use restricted in that module.

**Comment by tgnottingham** [comment]
I was hoping to reproduce the above perf results locally, but it's too much of a chore to figure out what's causing the difference. It's not that important anyway, provided it's not actually a correctness issue. I got some assurance that it isn't UB that's somehow causing the perf difference, by running some tests locally under Miri.

I pushed couple of changes: a fix to an existing bug in the signed LEB128 decoding (it had no perf impact, and may not have been encountered in practice), an increase to the `FileEncoder` buffer size (in case it helps performance), and modifications to address @cjgillot's reviews (thank you!).

Regarding performance, I'd like to see how #80115 changes things before trying anything else. Though if anyone can see something that will make buffered file writing more competitive with the previous implemention, I'd gladly take a hint.

**Comment by jyn514** [comment]
> Though if anyone can see something that will make buffered file writing more competitive with the previous implemention, I'd gladly take a hint.

Have you considered using mmap directly? That would need some unsafe code, but it shouldn't require an intermediate copy at all.

**Comment by tgnottingham** [comment]
> Have you considered using mmap directly? That would need some unsafe code, but it shouldn't require an intermediate copy at all.

I hadn't. I have some reservations, but it's worth experimenting with.

**Comment by tgnottingham** [comment]
I implemented an `mmap`-based version and tested it locally. I didn't optimize, because I just wanted to get a ballpark idea of the kind of improvement we could see. It would need to be significant to justify using `mmap`, IMO.

The numbers weren't very encouraging -- up to 7% worse in instruction count.

I'm sure it could result in an improvement if optimized, but I just don't want to go down that path, personally.

**Comment by tgnottingham** [comment]
I've temporarily integrated changes from #80115. I want to get an idea of the performance impact with that change, so I know whether or not this PR needs more perf work. Doing the perf run this way won't isolate the effects of just this PR, but that might be okay. If we see the stats are all still red, then more work will definitely need to be done.

**Comment by tgnottingham** [comment]
Okay, it will be pretty close after #80115. I'll consider this blocked on that.

**Comment by tgnottingham** [comment]
Found a micro-optimization. Can avoid updating the encoder position on every write by tracking position a little differently. I think that plus the blocking PR ought to do it.

File: compiler/rustc_serialize/src/opaque.rs [hunk 3]
```diff
@@ -141,10 +159,351 @@ impl serialize::Encoder for Encoder {
...
+                        *self.encoder_buffered = 0;
+                        *self.encoder_flushed += self.flushed;
+                    } else {
+                        let left = *self.encoder_buffered - self.flushed;
+
+                        unsafe {
+                            let src = self.buffer.as_ptr().add(self.flushed);
+                            let dst = self.buffer.as_mut_ptr();
+                            ptr::copy(src, dst, left);
+                        }
```
Comment (bjorn3) [inline compiler/rustc_serialize/src/opaque.rs:259]: It should be possible to use `split_at_mut` and `copy_from_slice` to make this safe.

Comment (tgnottingham) [reply compiler/rustc_serialize/src/opaque.rs]: Thanks! Ended up needing `copy_within`, since the ranges can overlap.

**Comment by tgnottingham** [comment]
Thanks for reviewing #80115, @oli-obk! Rebased.

**Comment by oli-obk** [comment]
Hmm... this is curious. It seems to affect a random subset of the queries `promoted_mir`, `generate_crate_metadata` `build_hir_map`, `eval_to_allocation_raw`, `optimized_mir` are all affected in some crates' `incremental-unchanged` test. Sometimes incremental loading has more executions?? Sometimes it just takes more time (or sometimes less) and sometimes the actual execution is slower. I can't really see a pattern, but that pattern may just be how often the fixed-size buffer is hit. Though I have no explanation for the execution count changes.

**Comment by tgnottingham** [comment]
@oli-obk, could you link to an example of the number of executions differing? I wasn't able to find one.

As for the time deltas, I don't think they're too surprising, if I understand correctly. Locally, I will see variance between two runs of the same compiler. And these time values are really small, so variance seems pretty likely.

I was surprised to see an instruction count improvement in the ctfe-stress-4 incr-unchanged benchmark. But that benchmark is special in that a large chunk of its cost is in serializing a single `u8` slice (in `rustc_middle::mir::interpret::Allocations`). The error-handling overhead of serializing that one slice with the new approach is fixed and tiny. I suspect the benefit of not having to reallocate and copy a potentially large vector buffer with the new approach becomes more evident in this benchmark.

**Comment by oli-obk** [comment]
> could you link to an example of the number of executions differing? I wasn't able to find one.

I misread the incremental loading columns, there are no execution count differences, and we do not have the concept of "loading count".

> As for the time deltas, I don't think they're too surprising, if I understand correctly. Locally, I will see variance between two runs of the same compiler. And these time values are really small, so variance seems pretty likely.

These values are not just plain noise, since the regression showed up in multiple runs. I agree with the non-suprising part though, the time values are very small, and thus small things can have fractional large effects.

Due to the fixed-size buffer, the regressed incremental runs may be creating a `1<<20` byte buffer even if they only use a fractional part of it. I wonder if incremental could also write the *used* buffer size to the cache and read that size for the next run as a serialization buffer size.

Though since it appears to be the actual serialization of specific queries that slowed down, maybe it's one of the other changes in this PR (like the bugfix). Could you try pulling out the test suite coverage increase and bugfixes and other "unrelated" changes into a separate PR or is it too much entangled? Or do you have any other ideas for how to progress? I do want this PR merged. It has small improvements to serialization speed for full runs (and some big ones on some edge cases like the ctfe stress tests), and if the regressions are due to bug fixes, then I see no reason to dig further.

**Comment by tgnottingham** [comment]
Sure, I'll separate out some unrelated changes. It would be good for this PR to shrink a bit.

The incremental regressions are almost certainly part of the core serialization change though. I spent a lot of time trying to reduce them, but eventually hit a bit of a wall, and felt like there would be better opportunities for improvements elsewhere. I got it down to the point where the memory trade-off seemed well worth it, though. Still, perhaps we can find an ideal solution...

The new approach has the advantage of never having to reallocate and grow its buffer, and much lower memory consumption, but the disadvantage of needing error handling on writes (creating the `Result`, checking its variant, and handling/propagating it). I don't think the advantages can be exploited further, so to eliminate the regressions, the amount or cost of error handling must be reduced.

Right now, error handling introduces overhead for each write of a primitive. It's possible to do better.

If we have a group of things we want to serialize (several primitives/objects, a struct, a collection, etc.), and we can reliably produce an upper bound on the encoded size, then we can do this:

```rust
if upper_bound <= self.buf.len() {
    if upper_bound > self.buf.len() - self.buffered {
        self.flush()?;
    }

    // Encode each thing directly into buffer without per-entity error handling.
    // Assuming that `upper_bound` is accurate, this won't overflow the buffer.
    ...
     
} else {
    // Fallback to writing each entity with per-entity error handling.
    ...
}
```

But I don't know how to do this in a way that is safe, performant, and clean.

We could also look into ways of making error handling itself cheaper. Perhaps `Result` creation codegen isn't optimal, for example.

**Comment by oli-obk** [comment]
Ok, thanks for elaborating. I think we should merge the PR then as it is, I can go through the changes commit by commit, so splitting them into PRs doesn't gain us too much.


Putting some of the numbers into the PR for future archeologists explaining why we are going to take the perf hit (it only really regresses if the test is in the millisecond range anyway, which I presume is caused by the static overhead of having a larger buffer and doing some more error checking):

The `rustc_incremental` change in bootstrapping looks like too much of a change for it to be noise, so that's great

> rustc_incremental | 18.865 | 10.396 | -44.9%

`ctfe-stress-4` has a 9% improvement on incremental-unchanged

`ucd`, `coercions` and `tuple-stress` have up to 1.7% regression on `incremental-unchanged` and really small regressions or small improvements on the other runs.

`max-rss` is green across the board with realistic tests like `clap`, `style-servo` and `webrender` having various improvements in the range of 10-30% :tada:

**Comment by tgnottingham** [comment]
Rebased to fix a conflict, squashed review commits, and added a comment on `FileEncoder`.

Also made one functional change, which was to return to using the default buffer size. I haven't seen any difference in performance in numerous local perf runs, so I don't think a larger buffer helps much.

**Comment by tgnottingham** [comment]
Instruction counts are worse after rebasing and changing the buffer size back to the default. I suspect it's just the rebase, but I'll increase the buffer size again to double check.

**Comment by tgnottingham** [comment]
Perf result confirms that larger buffer doesn't help.

**Comment by tgnottingham** [comment]
I'm looking into the overhead of error handling. From the assembly, it seems like there's unnecessary overhead in `Result` creation, but I could definitely be missing something. My assembly knowledge isn't great.

On first glance, it looks like creation of the `Ok` result is polluted with operations only relevant to creation of the `Err` result from the failed `flush()` path. But on further inspection, it looks like that work isn't useful for the `Err` result path, either.

Here is the assembly for `emit_u32`:

```asm
             push   %rbp
             push   %rbx
             push   %rax
             mov    %esi,%ebp
             mov    0x8(%rdi),%rbx
             mov    0x10(%rbx),%rax
             lea    0x5(%rax),%rcx
             cmp    0x8(%rbx),%rcx
   /-------- ja
/--|-------> mov    (%rbx),%rcx
|  |         add    %rax,%rcx
|  |         mov    $0x1,%edx
|  |         cmp    $0x80,%ebp
|  |  /----- jb
|  |  |      mov    %ebp,%esi
|  |  |  /-> mov    %esi,%edi
|  |  |  |   or     $0x80,%bpl
|  |  |  |   mov    %bpl,(%rcx)
|  |  |  |   shr    $0x7,%esi
|  |  |  |   add    $0x1,%rcx
|  |  |  |   add    $0x1,%rdx
|  |  |  |   mov    %esi,%ebp
|  |  |  |   cmp    $0x3fff,%edi
|  |  |  \-- ja
|  |  |      mov    %esi,%ebp
|  |  \----> mov    %bpl,(%rcx)
|  |         add    %rax,%rdx
|  |         mov    %rdx,0x10(%rbx)
|  |         mov    $0x3,%al          // Start of Result creation in Ok path
|  |  /----> shld   $0x8,%rcx,%rdx    // Continuation of Result creation in Ok and Err paths
|  |  |      shl    $0x8,%rcx
|  |  |      movzbl %al,%eax
|  |  |      or     %rcx,%rax
|  |  |      add    $0x8,%rsp
|  |  |      pop    %rbx
|  |  |      pop    %rbp
|  |  |      retq   
|  \--|----> mov    %rbx,%rdi
|     |      callq  *0x168146d(%rip) # <<rustc_serialize::opaque::FileEncoder>::flush@@Base+0xe7b158>
|     |      cmp    $0x3,%al
|     |  /-- jne
|     |  |   xor    %eax,%eax
\-----|--|-- jmp
      |  \-> mov    %rax,%rcx         // Start of Result creation in Err path (when flush() fails)
      |      shrd   $0x8,%rdx,%rcx
      |      shr    $0x8,%rdx
      \----- jmp
```

The result is a `Result<(), std::io::Error>`, which is 16 bytes. I'm under the impression the result is returned through `rax` and `rdx`.

The formation of the `Result` in the successful path:

```asm
|  |         mov    $0x3,%al        // Set Ok variant, I assume.
|  |  /----> shld   $0x8,%rcx,%rdx  // Seems irrelevant to Ok Result.
|  |  |      shl    $0x8,%rcx       // Seems irrelevant to Ok Result.
|  |  |      movzbl %al,%eax        // eax = Ok variant, zero-extended.
|  |  |      or     %rcx,%rax       // rax = Ok variant plus seemingly irrelevant stuff in rcx.
```

I think the above could be replaced with something like `mov $0x3,%rax`, and all the irrelevant junk moved to where it's relevant...except I don't think it's relevant to anything anyway, which leads me to the next part.

The formation of the `Result` in the unsuccessful flush path:

```asm
// ...just called flush, and found that result in rax was Err...

      |  \-> mov    %rax,%rcx
      |      shrd   $0x8,%rdx,%rcx
      |      shr    $0x8,%rdx
      \----- jmp

// ...jumps to the shld from previously quoted block...

|  |  /----> shld   $0x8,%rcx,%rdx
|  |  |      shl    $0x8,%rcx
|  |  |      movzbl %al,%eax
|  |  |      or     %rcx,%rax
```

IIUC, this sequence doesn't change `rax` and `rdx`, and the change to `rcx` isn't useful. And that would make sense, since this is the case where the `Err` result gets propagated. It's the same type, and I think the value is already in `rax` and `rdx`, so no work is necessary. I think all of this but the `jmp` could be removed, and its target changed to the instruction following the `or`.

If all the above is correct, fixing this would remove 4 instructions from the hot path, and 7 from the cold. That alone would help, since the hot code is really hot, but it might also unlock other optimizations. Anyway, I'm hoping my reasoning is right, and I'll continue looking into this.

**Comment by cjgillot** [comment]
Should we merge this PR for the memory improvement, and let the perf side for another PR?

**Comment by oli-obk** [comment]
> Should we merge this PR for the memory improvement, and let the perf side for another PR?

I'm fine with that, but if @tgnottingham still has motivation to munch away at the perf part, I'm not going to stop them :laughing:

**Comment by tgnottingham** [comment]
Rebased and returned to using default buffer size.

> Should we merge this PR for the memory improvement, and let the perf side for another PR?

I'd love that. :)

I'll create an issue for the `Result` problem (if I can prove there is one). I suspect it's an LLVM optimization issue, and it'll be a while before I can get up to speed on even the basics of LLVM.

**Comment by tgnottingham** [comment]
Created #81146 for the `Result` propagation performance issue.
